### PR TITLE
Spill of parallel aggregation in TiDB

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v7.6 (TiDB 7.6 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.4 (TiDB 7.4 versions)
- [ ] v7.3 (TiDB 7.3 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by XuHuaiyu — APPROVED** [review]

File: system-variables.md [hunk 1]
```diff
@@ -1688,6 +1688,19 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+> **警告：**
+>
+> 该变量控制的功能为实验特性，不建议在生产环境上使用，如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。
+
+- 作用域：SESSION | GLOBAL
+- 是否持久化到集群：是
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：布尔型
+- 默认值：`ON`
+- 这个变量用来控制是否支持并发 HashAgg 进行落盘。当该变量设置为 `true` 时，并发 HashAgg 将支持落盘。该变量将在功能正式发布时废弃。
```
Comment (qiancai) [inline system-variables.md:1702]:
```suggestion
- 这个变量用来控制 TiDB 是否支持并发 HashAgg 进行落盘。当该变量设置为 `true` 时，TiDB 支持并发 HashAgg 落盘。该变量将在功能正式发布时废弃。
```

File: system-variables.md [hunk 2]
```diff
@@ -1688,6 +1688,19 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
...
+
```
Comment (qiancai) [inline system-variables.md:1703]: 上面的 `true` 要修改为 `ON` 吗

Comment (xzhangxian1008) [reply system-variables.md:1703]: > 上面的 `true` 要修改为 `ON` 吗

好

File: configure-memory-usage.md [hunk 3]
```diff
@@ -148,8 +148,8 @@ TiDB 支持对执行算子的数据落盘功能。当 SQL 的内存使用超过
 
 - 落盘行为由参数 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)、[`tidb_enable_tmp_storage_on_oom`](/system-variables.md#tidb_enable_tmp_storage_on_oom)、[`tmp-storage-path`](/tidb-configuration-file.md#tmp-storage-path)、[`tmp-storage-quota`](/tidb-configuration-file.md#tmp-storage-quota) 共同控制。
 - 当落盘被触发时，TiDB 会在日志中打印一条包含关键字 `memory exceeds quota, spill to disk now` 或 `memory exceeds quota, set aggregate mode to spill-mode` 的日志。
-- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，HashAgg 落盘是从 v5.2.0 版本开始引入的。
-- 当包含 Sort、MergeJoin 或 HashJoin 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。当包含 HashAgg 算子的 SQL 语句引起内存 OOM 时，TiDB 默认不触发落盘，请设置系统变量 `tidb_executor_concurrency = 1` 来触发 HashAgg 落盘功能。
+- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，非并发 HashAgg 落盘是从 v5.2.0 版本开始引入的，并发 HashAgg 落盘是从 v7.6.0 版本开始引入的。
+- 当包含 Sort、MergeJoin 或 HashJoin 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。
```
Comment (qiancai) [inline configure-memory-usage.md:152]:
```suggestion
- 当包含 Sort、MergeJoin、HashJoin 或 HashAgg 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。
```

File: configure-memory-usage.md [hunk 4]
```diff
@@ -148,8 +148,8 @@ TiDB 支持对执行算子的数据落盘功能。当 SQL 的内存使用超过
 
 - 落盘行为由参数 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)、[`tidb_enable_tmp_storage_on_oom`](/system-variables.md#tidb_enable_tmp_storage_on_oom)、[`tmp-storage-path`](/tidb-configuration-file.md#tmp-storage-path)、[`tmp-storage-quota`](/tidb-configuration-file.md#tmp-storage-quota) 共同控制。
 - 当落盘被触发时，TiDB 会在日志中打印一条包含关键字 `memory exceeds quota, spill to disk now` 或 `memory exceeds quota, set aggregate mode to spill-mode` 的日志。
-- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，HashAgg 落盘是从 v5.2.0 版本开始引入的。
-- 当包含 Sort、MergeJoin 或 HashJoin 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。当包含 HashAgg 算子的 SQL 语句引起内存 OOM 时，TiDB 默认不触发落盘，请设置系统变量 `tidb_executor_concurrency = 1` 来触发 HashAgg 落盘功能。
+- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，从 v5.2.0 版本开始引入了参数 `tidb_executor_concurrency == 1` 的 HashAgg 算子的落盘，从 v7.6.0 版本开始引入了参数 `tidb_executor_concurrency > 1` 的 HashAgg 算子的落盘。
```
Comment (qiancai) [inline configure-memory-usage.md:151]:
```suggestion
- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的。非并发 HashAgg 算子的落盘是从 v5.2.0 版本开始引入的，需要设置系统变量 `tidb_executor_concurrency == 1` 来触发 HashAgg 落盘功能。并发 HashAgg 算子的落盘是从 v7.6.0 版本开始引入的，HashAgg 会在内存不足时自动尝试触发落盘。
```

**Review by Oreoxmt — APPROVED** [review]

File: system-variables.md [hunk 5]
```diff
@@ -1688,6 +1688,19 @@ mysql> SELECT job_info FROM mysql.analyze_jobs ORDER BY end_time DESC LIMIT 1;
 - 默认值：`OFF`
 - 这个变量用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集。关闭该变量后，之前收集的 `PREDICATE COLUMNS` 会被清除。详情见[收集部分列的统计信息](/statistics.md#收集部分列的统计信息)。
 
+### `tidb_enable_concurrent_hashagg_spill` <span class="version-mark">从 v7.6.0 版本开始引入</span>
```
Comment (qiancai) [inline system-variables.md:1691]:
```suggestion
### `tidb_enable_concurrent_hashagg_spill` <span class="version-mark">从 v8.0.0 版本开始引入</span>
```

File: configure-memory-usage.md [hunk 6]
```diff
@@ -148,8 +148,8 @@ TiDB 支持对执行算子的数据落盘功能。当 SQL 的内存使用超过
 
 - 落盘行为由参数 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)、[`tidb_enable_tmp_storage_on_oom`](/system-variables.md#tidb_enable_tmp_storage_on_oom)、[`tmp-storage-path`](/tidb-configuration-file.md#tmp-storage-path)、[`tmp-storage-quota`](/tidb-configuration-file.md#tmp-storage-quota) 共同控制。
 - 当落盘被触发时，TiDB 会在日志中打印一条包含关键字 `memory exceeds quota, spill to disk now` 或 `memory exceeds quota, set aggregate mode to spill-mode` 的日志。
-- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，HashAgg 落盘是从 v5.2.0 版本开始引入的。
-- 当包含 Sort、MergeJoin 或 HashJoin 的 SQL 语句引起内存 OOM 时，TiDB 默认会触发落盘。当包含 HashAgg 算子的 SQL 语句引起内存 OOM 时，TiDB 默认不触发落盘，请设置系统变量 `tidb_executor_concurrency = 1` 来触发 HashAgg 落盘功能。
+- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，非并发 HashAgg 的落盘是从 v5.2.0 版本开始引入的，并发 HashAgg 的落盘是从 v7.6.0 版本开始引入的。
```
Comment (qiancai) [inline configure-memory-usage.md:151]:
```suggestion
- Sort、MergeJoin、HashJoin 落盘是从 v4.0.0 版本开始引入的，非并发 HashAgg 的落盘是从 v5.2.0 版本开始引入的，并发 HashAgg 的落盘是从 v8.0.0 版本开始引入的。
```

**Review by qiancai — APPROVED** [review]

**Pull request merged**